<div style="text-align: center;">

# MA 203: Numerical Methods
## Tutorial-3

Shardul Junagade <br>
23110297

</div>

---
## **Question 1:**

The Stefan-Boltzmann law can be employed to estimate the rate of radiation of energy $ H $ from a surface, as in 

$$ H = A e \sigma T^{4} $$

where:
- $ H $ is in watts,
- $ A $ = the surface area ($ m^{2} $),
- $ e $ = the emissivity (dimensionless),
- $ \sigma $ = Stefan-Boltzmann constant ($ 5.67 \times 10^{-8} \, W m^{-2} K^{-4} $),
- $ T $ = absolute temperature ($ K $).

Determine the error of $ H $ for a steel plate with $ A = 0.15 \, m^{2} $, $ e = 0.90 $ and $ T = 650 \pm 20 \, K $. Compare your results with the exact error. Repeat the computation but with $ T = 650 \pm 40 $. Interpret your results.

### **Solution:**

![Solution 1](./images/tutorial3/1.jpg)

![Solution 1](./images/tutorial3/2.jpg)

![Solution 1](./images/tutorial3/3.jpg)

---
## **Question 2:**

Evaluate and interpret the condition numbers for:

a. $ \frac{\sin x}{1 + \cos x} $ for $ x = 1.0001\pi $ 

b. $ f(x) = \sqrt{x^2 + 1} - x $ for $ x = 300 $


### **Solution:**

![Solution 2](./images/tutorial3/4.jpg)

![Solution 2](./images/tutorial3/5.jpg)

---
## **Question 3:**

Real mechanical systems may involve the deflections of non-linear springs.  
In Figure 1 below, a mass $ m $ is released a distance $ h $ above a non-linear spring. The resistance force $ F $ of the spring is given by  

$$ F = -(k_1 d + k_2 d^{\frac{3}{2}}) $$

Conservation of energy can be used to show that  

$$ 0 = \frac{2k_2 d^{\frac{5}{2}}}{5} + \frac{1}{2} k_1 d^2 - mgd - mgh $$

<div style="text-align: center; width: 80%; margin: 0 auto;">
	<img src="./images/tutorial3/Figure1.png" alt="Figure 1">
</div>

Solve for $ d $ using the Bisection method, given the following parameter values: $ k_1 = 50,\!000 \, \frac{g}{s^2} $, $ k_2 = 40 \, \frac{g}{s^2 m^{0.5}} $, $ m = 90 \, g $, $ g = 9.81 \, \frac{m}{s^2} $, $ h = 0.45 \, m $.

### **Solution:**

The Bisection method is an iterative method that can be used to find the root of a function. The method requires an initial interval $ [a, b] $ such that $ f(a) \cdot f(b) < 0 $. The method then iteratively halves the interval and selects the subinterval that contains the root. The process is repeated until the interval is sufficiently small.

In this case, we are looking for the value of $ d $ that satisfies the equation:

$$ 0 = \frac{2k_2 d^{\frac{5}{2}}}{5} + \frac{1}{2} k_1 d^2 - mgd - mgh $$


In [ ]:
# Equation
def func(d, k1, k2, m, g, h):
    return (2*k2*d**(5/2))/5 + (1/2)*k1*d**2 - m*g*d - m*g*h

# Bisection method
def bisection_method(func, x_lower, x_upper, tol, k1, k2, m, g, h):		
    x_mid = (x_lower+x_upper)/2
    f_lower = func(x_lower, k1, k2, m, g, h)
    f_upper = func(x_upper, k1, k2, m, g, h)
    f_mid = func(x_mid, k1, k2, m, g, h)
    n = 0
    
	# Ensure the initial interval is valid
    if f_lower*f_upper > 0:
        raise ValueError("The function must have opposite signs at the interval bounds.")
    
    while abs(f_mid) > tol:
        if f_lower*f_mid < 0:
            x_upper = x_mid		# if the product of the function values at the lower and middle points is negative, the root is in the lower half of the interval
        else:
            x_lower = x_mid		# if the product of the function values at the upper and middle points is negative, the root is in the upper half of the interval
        
        x_mid = (x_lower+x_upper)/2
        f_lower = func(x_lower, k1, k2, m, g, h)
        f_upper = func(x_upper, k1, k2, m, g, h)
        f_mid = func(x_mid, k1, k2, m, g, h)
        n += 1
        
    return x_mid, f_mid, n

In [9]:
# Given parameters
k1 = 50000  # (N/m)
k2 = 40     # (N/m^(1.5))
m = 90      # mass (kg)
g = 9.81    # gravitational acceleration (m/s^2)
h = 0.45    # initial height (m)
tol = 1e-6  # tolerance for convergence

a = 0  # lower bound for d (m)
b = 1  # upper bound for d (m)

d, f, n = bisection_method(func, a, b, tol, k1, k2, m, g, h)
print(f"The solution for d in the equation is: {d:.8f} m")
print(f"The number of iterations is: {n}")
print(f"The final value of the function at the root is: {f:.14f}")

The solution for d in the equation is: 0.14493285 m
The number of iterations is: 29
The final value of the function at the root is: 0.00000083563953


The solution for $ d $ in the equation is approximately $ \textbf{0.14493285} \, m $. The number of iterations required for convergence is **29**, and the final value of the function at the root is approximately $ \textbf{0.00000083563953} $.

---
## **Question 4:**

The ideal gas equation of state is valid only for a limited range of pressures and temperatures. An alternative equation of state for gases is the van der Waals equation:

$$
\left( p + \frac{a}{v^2} \right) (v - b) = RT
$$

where $v$ is the molar volume, and $a$ and $b$ are the empirical constants for a gas. A chemical engineering design project requires you to accurately estimate the molar volume of ethyl alcohol ($a = 12.02 \, \text{L}^2 \cdot \text{atm} \cdot \text{mol}^{-2}$, $b = 0.08407 \, \text{L} \cdot \text{mol}^{-1}$) at a temperature of $T = 400 \, \text{K}$ and pressure of $p = 2.5 \, \text{atm}$. Use the false position method. Compare your results with the ideal gas law prediction $v_{\text{ideal}} = \frac{RT}{p}$.


### **Solution:**

The false position method is a numerical technique used to find the root of a function. It is an improved version of the bisection method, where instead of taking the midpoint, we use a weighted average to approximate the root more efficiently.

The van der Waals equation can be defined as a function of the molar volume $v$:
$$
f(v) = ( p + \frac{a}{v^2} ) (v - b) - RT = 0
$$

We need to find the molar volume $v$ that satisfies this equation. 

The false position method requires two initial guesses, $v_0$ and $v_1$, such that $f(v_0)$ and $f(v_1)$ have opposite signs. The next guess, $v_2$, is calculated as:
$$
v_2 = v_1 - f(v_1) \frac{v_1 - v_0}{f(v_1) - f(v_0)}
$$
The process is repeated until the function value is within a specified tolerance. 



In [ ]:
# Vander Waals Equation
def f(v, p, T, R, a, b):
    return (p + (a / (v ** 2))) * (v - b) - (R * T)

# False Position Method
def false_position(f, x0, x1, tol=1e-6, max_iter=100):
    if f(x0) * f(x1) >= 0:
        print("Error: No root in the initial bracket. Try different values.")
        return None

    for i in range(max_iter):
        x2 = x1 - (f(x1) * (x1 - x0)) / (f(x1) - f(x0)) # formula
        print(f"Iteration {i+1}: x2 = {x2:.6f}, f(x2) = {f(x2):.6f}")

        if abs(f(x2)) < tol:
            return x2, i+1

        if f(x0)*f(x2) < 0:
            x1 = x2
        else:
            x0 = x2

    return x2, i+1


# Constants
R = 0.08206  	# L.atm/(K.mol)
a = 12.02  		# L^2.atm/mol^2
b = 0.08407 	# L/mol
T = 400  		# K
p = 2.5  		# atm
tol = 1e-6		# Tolerance

# Initial guesses
x0 = 10
x1 = 15

vanderwaal_v, iterations = false_position(lambda v: f(v, p, T, R, a, b), x0, x1, tol)
print(f"\nVolume according to Van der Waals equation = {vanderwaal_v:.6f} L/mol")

ideal_v = R*T / p
print(f"Volume according to ideal gas law = {ideal_v:.6f} L/mol")

print(f"Difference = {abs(ideal_v - vanderwaal_v):.6f} L/mol")
print(f"Percentage error = {(abs(ideal_v - vanderwaal_v) / ideal_v * 100):.2f}%")
print("Number of iterations:", iterations)

Iteration 1: x2 = 12.826233, f(x2) = -0.037593
Iteration 2: x2 = 12.841651, f(x2) = -0.000159
Iteration 3: x2 = 12.841716, f(x2) = -0.000001

Volume according to Van der Waals equation = 12.841716 L/mol
Volume according to ideal gas law = 13.129600 L/mol
Difference = 0.287884 L/mol
Percentage error = 2.19%
Number of iterations: 3


**Results:**

- The molar volume of ethyl alcohol at $T = 400 \, \text{K}$ and $p = 2.5 \, \text{atm}$ is approximately $12.841716 \, \text{L/mol}$ using the van der Waals equation of state. 
- The ideal gas law prediction gives a molar volume of $13.129600 \, \text{L/mol}$. 
- The absolute difference is $0.287884 \, \text{L/mol}$, which is a percentage difference of $2.19\%$. 
- The number of iterations required was $3$.
